src/rust/src/backend/keys.rs

```diff
@@ -64,13 +65,110 @@ fn load_pem_private_key<'p>(
     unsafe_skip_rsa_key_validation: bool,
 ) -> CryptographyResult<pyo3::Bound<'p, pyo3::PyAny>> {
     let _ = backend;
-    let password = password.as_ref().map(CffiBuf::as_bytes);
-    let mut status = utils::PasswordCallbackStatus::Unused;
-    let pkey = openssl::pkey::PKey::private_key_from_pem_callback(
+
+    let p = x509::find_in_pem(
         data.as_bytes(),
-        utils::password_callback(&mut status, password),
-    );
-    let pkey = utils::handle_key_load_result(py, pkey, status, password)?;
+        |p| ["PRIVATE KEY", "ENCRYPTED PRIVATE KEY", "RSA PRIVATE KEY", "EC PRIVATE KEY", "DSA PRIVATE KEY"].contains(&p.tag()),
+        "Valid PEM but no BEGIN/END delimiters for a private key found. Are you sure this is a private key?"
+    )?;
+    let password = password.as_ref().map(|v| v.as_bytes());
+    let mut password_used = false;
+    // TODO: Surely we can avoid this clone?
+    let tag = p.tag().to_string();
+    let data = match p.headers().get("Proc-Type") {
+        Some("4,ENCRYPTED") => {
+            password_used = true;
+            let Some(dek_info) = p.headers().get("DEK-Info") else {
+                return Err(CryptographyError::from(
+                    pyo3::exceptions::PyValueError::new_err(
+                        "Encrypted PEM doesn't have a DEK-Info header.",
+                    ),
+                ));
+            };
+            let Some((cipher_algorithm, iv)) = dek_info.split_once(',') else {
+                return Err(CryptographyError::from(
+                    pyo3::exceptions::PyValueError::new_err(
+                        "Encrypted PEM's DEK-Info header is not valid.",
+                    ),
+                ));
+            };
+
+            let password = match password {
+                None | Some(b"") => {
+                    return Err(CryptographyError::from(
+                        pyo3::exceptions::PyTypeError::new_err(
+                            "Password was not given but private key is encrypted",
+                        ),
+                    ))
+                }
+                Some(p) => p,
+            };
+
+            let cipher = match cipher_algorithm {
```

Comment:
Could we put a comment telling future readers why we only support these 3? (Just putting this here in addition to our chat)